### PR TITLE
Steady state

### PR DESCRIPTION
**Pull request by PavelBal**
Adds Steady state calculation/export (compare discussions in https://github.com/Open-Systems-Pharmacology/Forum/discussions/1450) originally implemented in `esqlabsR`.

We also have a function to export such a "steady state" to excel file that can be loaded in MoBi, would it be also of interest for the main package?

```
#' Export steady-state to excel
#'
#' @details Simulates a given model to its steady-state and creates an
#' Excel-file with the end values of molecules amounts in all containers and
#' parameter values that have a right-hand-side (state variable parameters).
#' @param resultsXLSPath Path to the xls-file where the results will be written
#'   to. If the file does not exist, a new file is created. If no path is
#'   provided, the file will be created in the same directory where the model
#'   file is located. The name of the file will be `<SimulationFileName>_SS`.
#' @param simulation A `Simulation` object that will be updated with the steady
#' state
#' @inheritParams ospsuite.parameteridentification::getSteadyState
#' @import ospsuite.utils ospsuite.parameteridentification
#' @export
exportSteadyStateToXLS <- function(simulation,
                                   quantitiesPaths = NULL,
                                   resultsXLSPath = "",
                                   steadyStateTime = 1000,
                                   ignoreIfFormula = TRUE,
                                   stopIfNotFound = TRUE,
                                   lowerThreshold = 1e-15,
                                   simulationRunOptions = NULL)
```

### CODE REVIEW

File: R/utilities-simulation.R
Comment (Yuri05) [inline R/utilities-simulation.R:800]: I would change the order of arguments: first `simulations`, then `quantitiesPaths `

**Review by Felixmil — DISMISSED** [review]
Approved once `pkgdown::build_site()` is run

File: R/utilities-simulation.R
Comment (Yuri05) [inline R/utilities-simulation.R:802]: Should we try to select the default simulation time based on the start time of the last application?
Instead of setting  the default value to 1000: set it NULL.
And then estimate the default simulation time as Time-of-last-application + Delta
Where Delta = 1 day or so (I personally would set it even to 3 days)

File: R/utilities-simulation.R
Comment (Yuri05) [inline R/utilities-simulation.R:843]: It would be nice to have the function `setOutputs(quantityPaths)` in the ospsuite-R, which would just execute the 2 functions above (clearOutputs and addOutputs )

File: R/utilities-simulation.R
Comment (Yuri05) [inline R/utilities-simulation.R:804]: In the Ospsuite.Core we calculate a Comparison Threshold for each state variable and each observer dynamically, based on the AbsTol and ScaleFactor for the state variables and on the monitor formula for observers; also having special handling for Fractions.
Values below that comparison tolerance are considered numerical noise.
I wonder if we should export this property to R and use as a default for each variable if not passed by user

File: R/utilities-simulation.R
Comment (Yuri05) [inline R/utilities-simulation.R:777]: I don't get this: how can the values for SOME species be not available if the simulation run was successful?

File: R/utilities-simulation.R
Comment (Yuri05) [inline R/utilities-simulation.R:780]: Strictly speaking, we should also check if the variable allows negative values and replace not the values below the threshold by 0 but only the values in the interval [-Threshold..Threshold]

File: R/utilities-simulation.R
Comment (Yuri05) [inline R/utilities-simulation.R:875]: I don't get this: how is a state variable supposed to be defined by explicit formula???

File: R/utilities-simulation.R
Comment (PavelBal) [reply R/utilities-simulation.R:802]: Hmm, I don't see an easy way to do this. How would you get all applications? There is a method `allApplicationsFor()`, but then we need the list of all molecules first, and then retrieve all applications and compare... I don't think the complexity is worth it.

What I can imagin for the future, is to run simulation until the derivatives become near 0.

Comment (Yuri05) [reply R/utilities-simulation.R:802]: Get all molecules; for each molecule get all applications; for each application get its start time.
Get max of all start times.
```
> sim=loadSimulation("c:/Temp/S_C1C2.pkml")
> sim$allXenobioticFloatingMoleculeNames()
[1] "C1" "C2"
> sim$allApplicationsFor("C1")
[[1]]
Application: 
   Start time: 0.00e+00 [min] 
   Infusion time: 1.00 [min] 
   DrugMass: 182.50 [µmol] 

[[2]]
Application: 
   Start time: 720.00 [min] 
   Infusion time: 1.00 [min] 
   DrugMass: 182.50 [µmol] 

> sim$allApplicationsFor("C2")
[[1]]
Application: 
   Start time: 0.00e+00 [min] 
   Infusion time: 1.00 [min] 
   DrugMass: 182.50 [µmol] 

[[2]]
Application: 
   Start time: 480.00 [min] 
   Infusion time: 1.00 [min] 
   DrugMass: 182.50 [µmol] 
```

File: R/utilities-simulation.R
Comment (PavelBal) [reply R/utilities-simulation.R:843]: PR created

File: R/utilities-simulation.R
Comment (PavelBal) [reply R/utilities-simulation.R:777]: If I remember correctly, it was some observer that was `NaN` the whole time... there was a case for this.

File: R/utilities-simulation.R
Comment (PavelBal) [reply R/utilities-simulation.R:875]: What we are dealing here with is the **start value** of the parameter, which can be an explicit formula.

File: R/utilities-simulation.R
Comment (PavelBal) [reply R/utilities-simulation.R:780]: You are right. For this, https://github.com/Open-Systems-Pharmacology/OSPSuite-R/issues/307 must be implemented.

File: R/utilities-simulation.R
Comment (Yuri05) [reply R/utilities-simulation.R:777]: or, so with "species" you mean also observers? (I thought only ODE state variables were meant here).
Then this could happen indeed.

File: R/utilities-simulation.R
Comment (Yuri05) [reply R/utilities-simulation.R:875]: ok... but I still don't understand the use case. So if a parameter is a state variable with initial value defined by formula - why would I need to ignore this parameter in the steady state?

File: R/utilities-simulation.R
Comment (Yuri05) [reply R/utilities-simulation.R:780]: Issue create: https://github.com/Open-Systems-Pharmacology/OSPSuite-R/issues/1348

File: R/utilities-simulation.R
Comment (PavelBal) [reply R/utilities-simulation.R]: Done.

File: R/utilities-simulation.R
Comment (PavelBal) [reply R/utilities-simulation.R:807]: I am not sure this works properly... This is what I wanted to avoid:

image

Amount of an orally administered drug in stomach.

File: R/utilities-simulation.R
Comment (PavelBal) [reply R/utilities-simulation.R:903]: I sometimes define the start value of the parameter as a formula that calculates the steady-state of this parameter based on other intputs. In this case, I would not like to overwrite the start value.

File: R/utilities-simulation.R
Comment (Yuri05) [reply R/utilities-simulation.R:807]: Hmm, it's theoretically done in SimModel:
https://github.com/Open-Systems-Pharmacology/OSPSuite.SimModel/blob/7092dd785819f9904abbc1f40cc33c8e4be05a30/src/OSPSuite.SimModelNative/src/DESolver.cpp#L350-L364

However when looking into the implementation: only the values in `[-Threshold .. 0]` are set to zero, but **not the values in** `[0 .. Threshold]`
https://github.com/Open-Systems-Pharmacology/OSPSuite.SimModel/blob/7092dd785819f9904abbc1f40cc33c8e4be05a30/src/OSPSuite.SimModelNative/src/SimulationTask.cpp#L186C1-L193C1

I assume originally the values in [-Threshold .. Threshold] were set to zero, but we got some kind of problems with this (but I don't remember which exactly..). Will create an issue for this in SimModel

Comment (Yuri05) [reply R/utilities-simulation.R:807]: s. https://github.com/Open-Systems-Pharmacology/OSPSuite.SimModel/issues/162

File: R/utilities-simulation.R
Comment (Yuri05) [reply R/utilities-simulation.R:903]: While I still don't understand the use case (maybe you can explain it to me during one of the meetings) - I wonder if we should set the default value of `ignoreIfFormula `to **FALSE**?
It seems pretty dangerous for me when the steady state of an ODE species is ignored because its initial value is not const. but defined by the formula.

Comment (PavelBal) [reply R/utilities-simulation.R:903]: I have been thinking about it for some time now, and I am still not convinced, maybe we should discuss it in a meeting.

The way I see it, if a start value of an ODE parameter (and the same goes for molecules, e.g. start values of proteins) is defined by a formula, then this formula represents the steady-state. Otherwise, there is something off with the equation, don't you think?

**Comment by Yuri05** [comment]
@PavelBal Let's create issues for:
* https://github.com/Open-Systems-Pharmacology/OSPSuite-R/pull/1341#pullrequestreview-1794487002
* https://github.com/Open-Systems-Pharmacology/OSPSuite-R/pull/1341#pullrequestreview-1794496717
* https://github.com/Open-Systems-Pharmacology/OSPSuite-R/pull/1341#pullrequestreview-1794510622
After that I would say we can merge this PR and address those issues in further PRs if required

**Review by Yuri05 — APPROVED** [review]

**Comment by PavelBal** [comment]
@Yuri05 Can you re-review? I have removed the `stopIfNotFound` argument and had to update the snapshots to match the latest TLF version.

File: R/utilities-simulation.R
Comment (Yuri05) [inline R/utilities-simulation.R:797]: Is there also `setSteadyState()` or `setInitialState()` function somewhere?

**Review by Yuri05 — APPROVED** [review]

File: R/utilities-simulation.R
Comment (PavelBal) [reply R/utilities-simulation.R:797]: No, the workflow would be to call

```
#' setQuantityValuesByPath(
#'   quantityPaths = steadyState[[sim$id]]$paths,
#'   values = steadyState[[sim$id]]$values, simulation = sim
#' )
```

as described in the example of `getSteadyState()`. Unless we want to have additional checks or whatever, I do not see a need for a separate function, be we can implement something if required.

**Pull request merged**
